projects/real_world_ovmm/build_3d_map.py

```diff
@@ -194,35 +203,72 @@ def run_exploration(
     for i in range(explore_iter):
         print("\n" * 2)
         print("-" * 20, i, "-" * 20)
-        # sample a goal
-        goal = space.sample_frontier().cpu().numpy()
-        if goal is None:
-            print(" ------ sampling failed!")
         start = robot.get_base_pose()
+        start_is_valid = space.is_valid(start)
+        print("Start is valid:", start_is_valid)
+        # if start is not valid move backwards a bit
+        if not start_is_valid:
+            print("Start not valid. back up a bit.")
+            robot.nav.navigate_to([-0.1, 0, 0], relative=True)
+            continue
         print("       Start:", start)
-        print("Sampled Goal:", goal)
-        show_goal = np.zeros(3)
-        show_goal[:2] = goal[:2]
-        print("Start is valid:", collector.voxel_map.xyt_is_safe(start))
-        print(" Goal is valid:", collector.voxel_map.xyt_is_safe(goal))
-        # plan to the sampled goal
-        res = planner.plan(start, goal)
-        print("Found plan:", res.success)
-        obstacles, explored = collector.get_2d_map()
-        img = (10 * obstacles) + explored
-        space.draw_state_on_grid(img, start, weight=5)
-        space.draw_state_on_grid(img, goal, weight=5)
-        plt.imshow(img)
-        if res.success:
-            path = collector.voxel_map.plan_to_grid_coords(res)
-            x, y = get_x_and_y_from_path(path)
-            plt.plot(y, x)
+        # sample a goal
+        if random_goals:
+            goal = next(space.sample_random_frontier()).cpu().numpy()
         else:
-            continue
+            # extract goal using fmm planner
+            tries = 0
+            failed = False
+            for goal in space.sample_closest_frontier(start):
+                if goal is None:
+                    failed = True
+                    break
+                goal = goal.cpu().numpy()
+                print("Sampled Goal:", goal)
+                show_goal = np.zeros(3)
+                show_goal[:2] = goal[:2]
+                goal_is_valid = space.is_valid(goal)
+                print("Start is valid:", start_is_valid)
+                print(" Goal is valid:", goal_is_valid)
+                if not goal_is_valid:
+                    print(" -> resample goal.")
+                    continue
+                # plan to the sampled goal
+                res = planner.plan(start, goal)
+                print("Found plan:", res.success)
+                obstacles, explored = collector.get_2d_map()
+                img = (10 * obstacles) + explored
+                space.draw_state_on_grid(img, start, weight=5)
+                space.draw_state_on_grid(img, goal, weight=5)
```

Comment:
All this can go inside the visualize if condition, if we're not visualizing there's no need to compute it